core/src/main/scala/org/apache/spark/sql/rapids/tool/util/ToolsPlanGraph.scala

```diff
@@ -16,16 +16,410 @@
+    // First normalize the name.
+    val normalizedName = ToolsPlanGraph.processPlanInfo(node.name)
+    val stageIds = getNodeStagesByAccum(node)
+    normalizedName match {
+      case nName if isEpilogueExec(nName) =>
+        // cases that are supposed to be tail of the stage cluster
+        // only use the smallest StageId for now
+        if (stageIds.size <= 1) {
+          stageIds
+        } else {
+          Set[Int](stageIds.min)
+        }
+      case nName if isPrologueExec(nName) =>
+        // cases that are supposed to be head of a new stage. We should pick the stages associated
+        // with the reading metrics (pick the highest number for simplicity).
+        if (stageIds.size <= 1) {
+          ToolsPlanGraph.EMPTY_CLUSTERS
+        } else {
+          Set[Int](stageIds.max)
+        }
+      case _ =>
+        // Everything else should be handled here. If a node a has more than one stage,
+        // we should handle that here
+        //TODO: assert(stageIds.size <= 1)
+        //stageIds.headOption
+        stageIds
+    }
+  }
+
+  /**
+   * Updates the data structure that keeps track of the nodes cluster assignment.
+   * It adds the node to the map and remove the node from the orphans list if it exists.
+   * @param node the node to be assigned
+   * @param orphanNodes the list of nodes that are not assigned to any cluster
+   * @param clusters the clusterIds to assign the node to
+   */
+  private def removeNodeFromOrphans(node: SparkPlanGraphNode,
+      orphanNodes: mutable.ArrayBuffer[SparkPlanGraphNode],
+      clusters: Set[Int]): Unit = {
+    nodeToStageCluster.put(node.id, clusters)
+    orphanNodes -= node
+  }
+
+  /**
+   * Commits a wholeStageNode to a cluster.
+   * A WholeStageNode is visited after its children are. If any of the children is not assigned to
+   * a cluster, the wNode will transfer its assignment to the child.
+   * @param wNode the wholeStageCodeGen node to be visited
+   * @param orphanNodes the list of nodes that are not assigned to any cluster.
+   * @param clusters the clusterId to assign the node to.
+   * @return true if a change is made.
+   */
+  private def commitNodeToStageCluster(
+    wNode: SparkPlanGraphCluster,
+    orphanNodes: mutable.ArrayBuffer[SparkPlanGraphNode],
+    clusters: Set[Int]): Boolean = {
+    if (nodeToStageCluster.contains(wNode.id) && clusters.subsetOf(nodeToStageCluster(wNode.id))) {
+      // Node is assigned to the same cluster before. Nothing to be done.
+      false
+    } else {
+      val newClusterIds =
+        clusters ++ nodeToStageCluster.getOrElse(wNode.id, ToolsPlanGraph.EMPTY_CLUSTERS)
+      // remove the wNode from orphanNodes if it exists
+      removeNodeFromOrphans(wNode, orphanNodes, newClusterIds)
+      // assign the children to the same clusters if any of them is not assigned already.
+      wNode.nodes.foreach { childNode =>
+        // TODO assert that the child node is not assigned something different than the wStage
```

Comment:
Can we add the assert mentioned in the TODO.

Reply:
I removed the assert statement from the implementation as we don't want the code to break.